### PR TITLE
feat(GraphQL support): RFC for GraphQL support in datahub-frontend (Part 2/2) - Mutations

### PR DESCRIPTION
**Pull request by jjoyce0510**
Pre-read: [RFC Part 1/2](https://github.com/linkedin/datahub/pulls)

This PR is part 2/2 of the GraphQL in `datahub-frontend` proposal, outlining the changes required to support writes (mutations) against the metadata catalog. 

The RFC goes into depth on specific modeling considerations, proposing the omission of the entity-aspect distinction at the API layer. 

Additionally, the RFC outlines how we can get to functional parity with the existing Play API, covering:
- authentication 
- authorization
- search + autocomplete
- browse + browse path

as modeled in GraphQL. 

I've created a reference branch to showcase the changes outlined in this document. A PR from that branch can be viewed [here](https://github.com/linkedin/datahub/pull/2044). Notice that the branch has evolved to contain both changes related to GraphQL Queries (RFC 1/2) and Mutations (RFC 2/2)

Notice that one thing I'm still looking for is an easy way to auto-generated POJOs from the GraphQL spec to making writing resolvers a more strongly typed experience. Coming soon (hopefully) -- Stay tuned!

Fully-rendered version [here](https://github.com/jjoyce0510/datahub-1/blob/d2651a790078535754e95157006d6f4127a94eba/docs/rfc/active/graph_ql_frontend/mutations.md). 

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: docs/rfc/active/graph_ql_frontend/mutations.md [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+**System** types include special "root" types which provide entry points to the graph:
+- `Query`: Reads against the graph (covered in RFC 2042)
+- `Mutation`: Writes against the graph 
+- `Subscription`: Subscribing to changes within the graph
+
+In this design, we focus on extending our GQL schema to model `Mutations`. Let's start by looking at how we can model mutations.
+
+**The Approach**
+
+Continuing the theme of **entity-oriented** client side operations, we propose limiting mutation to the entity level. The avoids leaking the GMA entity-aspect distinction into the client side, simplifying the model for frontend development and minimizing the effort required to evolve GMA entities.
```
Comment (shirshanka) [inline docs/rfc/active/graph_ql_frontend/mutations.md:56]: +1 for entity-oriented

**Comment by shirshanka** [comment]
Went through it. This looks great. Thanks @jjoyce0510!

As you have already identified, automated generation of the GraphQL types from PDL is key to keeping the "manual edit burden" to a minimum. 

Excited to see this moving forward and looking forward to feedback from others.

File: docs/rfc/active/graph_ql_frontend/mutations.md [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+
+
+## Alternatives
+### Alternatives to GQL
+
+Keep the resource-oriented approach, with different endpoints for each aspect needing updated. 
+
+### Alternate GQL Modeling Convention 
+
+*Mutation-per-aspect*
```
Comment (mars-lan) [inline docs/rfc/active/graph_ql_frontend/mutations.md:545]: Let's put the concept of aspects aside, it's not uncommon to have mutations that do partial update of an entity/type in GraphQL, e.g. `addCommentToPost`, `markItemComplete` etc. If the ultimate goal is to code-gen these, I feel that having an aspect-specific mutation seems like a better approach?

Comment (jjoyce0510) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: Hmm the types of mutations you're describing seem useful in simplifying the mutation experience of _collection_ fields. Let's say you want to add an entry to an existing map or array, change a specific item, or delete an element.  In such cases it may be a better experience for the server to handle these changes to reduce complexity on the client. 

It may be worthwhile to consider auto-generating such mutations for nested _collection_ fields, but I don't see the value in having these on a per-aspect basis. But maybe I'm overlooking something; do you see concrete value in exposing the aspect abstraction to the frontend, keeping the aspect mutation path separate from top-level entity mutation?

Comment (mars-lan) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: A few more points to consider

1. Not every aspect is/should be editable by the frontend. In fact, most of the aspects are populated via MCE only
2. The entity-oriented mutation creates the confusion that partial update is fully supported at all nesting levels (even within aspects), whereas in reality, it's only the first level (aspect-level)

Unfortunately the whole GraphQL mutation thing is kind of an afterthought and lacks a good structure to it. It seems that in general people create mutations for specific purposes rather than mirroring the query structure.

Comment (jjoyce0510) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: 1. Totally acknowledge this point. GraphQL requires separate types for query and mutation. I'm thinking that the mutation "input" types will only include those fields which are editable via the rest API. Once we get to auto-generation, this can come from annotated "read-only" fields in the API value model (eg. Dataset.pdl) 

2. The partial update _should_ also apply at the nested level, so long as fields in those nested models are "optional" in the GQL input types. If they are required but not provided the mutation validation will fail. The case where "partial update" doesn't really apply is in lists, regardless of where they appear (entity or aspect level), which will always imply a full replacement of the existing value...

File: docs/rfc/active/graph_ql_frontend/mutations.md [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+
+*Folks from DH have noted that the entity "ingest" API is intended to be private, only used by the MCE consumer logic. We propose formal support for this API (or something of equivalent function) for atomically updating multiple entity aspects at one time.
+
+*A related point worth considering is the modeling inconsistency in APIs exposed by GMA. Today, we are able to *read* entity models like `Dataset.pdl` but are unable to *write* using this model. We should aim for clarity & consistency in the model exposed to clients of GMA by exposing `Entity` or `Snapshot` models for both read *&* write -- not a mixture of both. To work around this inconsistency, the sample code provided with this RFC includes entity-to-snapshot conversion hidden within the `Datasets` DAO, thereby allowing the API layer to concern itself only with the canonical entity model.
```
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:214]: In either case I think this API should be considerate of how it would look if was the replacement for restli, so that we can easily replace restli with it in the future. The resolvers, rather than read/write via the restli APIs, would just read/write to DBs. That migration be transparent to anyone using GQL before the migration, if we can keep the APIs the same.

File: docs/rfc/active/graph_ql_frontend/mutations.md [hunk 4]
```diff
@@ -0,0 +1,647 @@
...
+    browse(input: BrowseInput!): BrowseResults
+}
+```
+
+which accepts a`BrowseInput`
+
+```graphql
+input BrowseInput {
+	type: EntityType!
+	path: String
```
Comment (jplaisted) [inline docs/rfc/active/graph_ql_frontend/mutations.md:478]: Would this make more sense as `pathParts: String[]` so that encoding of paths is not needed?

File: docs/rfc/active/graph_ql_frontend/mutations.md [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+- input fields aside from the primary key (urn) are *optional*, meaning partial update is permitted
+- update / delete mutations are modeled at the `Dataset` level. The `DatasetUpdateInput` represents a "patch" object to be upserted into GMA.
+
+Using this modeling approach, we can choose to include or exclude specific entity aspects that we'd like to expose for mutation. For example, if we don't want to allow changes to a `Dataset`'s schema metadata, we can omit inclusion of an `input` object representing the `Schema` aspect.
+
+This modeling convention is
+
+1. **Simple**: Continuing the convention introduced on the read side, we want to keep the API logically simple by orienting operations around top-level entities, as opposed to co-locating operations against entities and aspects. Keeping the model hierarchical simplifies the frontend development model. 
+
+2. **Easily Extensible**: Adding support for writes against new aspects means adding a single new `input` type, and implementing execution logic (detailed in subsequent sections). There's no need to introduce a new mutation type.
```
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:109]: I ack this was touched on at the end of the RFC, but "easy" seems subjective here. Easy to modify files, sure. But is that really easy for the long term to customize different deployments? Probably not, imo.

File: docs/rfc/active/graph_ql_frontend/mutations.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:214]: Maybe we should have both aspects and entities be readable / writable?

The entity code would use GQL to read/write the aspect GQL, which turn turn would read/write to the restli endpoints (though this also requires every aspect have a restli endpoint, which I don't think is true today?), or in the future right to the DB. So we could have the entity layer "on top of" the aspect layer. While this doesn't get you distributed extensions, it does more closely reflect the underlying data model, and what we have today.

Not sure if I like it or not, just a thought.

File: docs/rfc/active/graph_ql_frontend/mutations.md (hunk 4 above)
Comment (jjoyce0510) [reply docs/rfc/active/graph_ql_frontend/mutations.md:478]: This should be better :)

Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:478]: Any particular reason why?

Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:478]: I'm generally not a fan of including string fields that are up for interpretation. Requires encoders and decoders, and often people don't make those but just do it in line and then the logic isn't reusable and is unclear and meaning gets lost to time. /rant

tl;dr avoid stringly typed code if possible

File: docs/rfc/active/graph_ql_frontend/mutations.md (hunk 2 above)
Comment (mars-lan) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: 2. Currently the `ingest` endpoint doesn't support partial update so is the plan to add that?

Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: +1, we'd need to add that support to GMSes / GMA, which may be non trivial.

Comment (mars-lan) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: Also the semantics seems kind of strange around `required` fields when it comes to partial update, i.e. one must know its current value in order to do a partial update of other optional fields.

Comment (jplaisted) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: That's... another good point. Partial update might just need to take a schema-less json blob as a result :/

Proto got rid of `required` and everything is implicitly `optional` as of proto3. Yet another interesting piece of evidence that supports it.

Comment (jjoyce0510) [reply docs/rfc/active/graph_ql_frontend/mutations.md:545]: You guys bring up great points. Since GMS does not currently support partial aspect updates, we'd be left with the following options to achieve an entity-oriented approach:

1. Require all aspect fields on input types --> As Mars mentioned early, this means that partial update semantics will only apply at the entity level and not the aspect level, where we will in effect be limited to 'creates'. To be fair, this is true for the the existing 'create' subresources in GMS as well (eg. Dataset Ownership)

2. Implement support for 'partial' aspect updates in GMS. Flow at GMS would be: read latest version of aspect, apply patch, write back. 

To address the point about `required` fields -- if I remember correctly, Rest.li Patch semantics are such that the incoming patch is _not_ validated against the full schema (a partial schema is constructed). Instead, it is the intention that the record resulting from _applying_ a patch be validated. From the Rest.li [docs](https://linkedin.github.io/rest.li/Validation-in-Rest_li#specifying-validation-rules):

> For partial update requests (patches), the goal is to ensure that if the patch is applied to a valid entity, the modified entity will also be valid. For example, a patch that deletes a required field is invalid because the modified entity will be missing that field

Either way, my goal is to provide aspect operations that are opaque to the end client. A nice byproduct is that entity GMS API models can become consistent across reads and writes (eg. Dataset.pdl for everything). 

Let's continue to think on this, perhaps Mutation implementation can be decoupled from the Query implementation until we have alignment on guidance here.

**Comment by jjoyce0510** [comment]
Closing this as it was merged in a long time ago.
